sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -421,46 +439,75 @@ object ScalaReflection extends ScalaReflection {
     if (!inputObject.dataType.isInstanceOf[ObjectType]) {
       inputObject
     } else {
+      val className = getClassNameFromType(tpe)
       tpe match {
         case t if t <:< localTypeOf[Option[_]] =>
           val TypeRef(_, _, Seq(optType)) = t
           optType match {
             // For primitive types we must manually unbox the value of the object.
             case t if t <:< definitions.IntTpe =>
-              Invoke(
-                UnwrapOption(ObjectType(classOf[java.lang.Integer]), inputObject),
-                "intValue",
-                IntegerType)
+              val unwrapped = UnwrapOption(ObjectType(classOf[java.lang.Integer]), inputObject)
+              expressions.If(
+                IsNull(unwrapped),
+                expressions.Literal.create(null, silentSchemaFor(optType).dataType),
+                Invoke(
+                  unwrapped,
```

Comment:
I think you are right. I add this to make sure it is not possibly creating wrong values when I debug. I will remove these `If`.